### PR TITLE
[WIP] Fix #105

### LINKED ISSUE #105: Deleting a contribution does not release the slug for further use.
softdelete hook triggers but it keeps the slug for the contribution "reserved" - hence a contribution with the same slug cannot be created ever again. 

Possible solution: randomize slug on softdelete or add a number to slug on create if exists.

### PR DESCRIPTION
**Pull request by roschaefer**
I would like to write a unit test for this before we merge..

Close #105

### CODE REVIEW

**Comment by sebako86** [comment]
sure thing. Let me know when the test is ready 👍

**Comment by roschaefer** [comment]
This PR is blocked as many unrelated tests fail because of:
```
   users patch
(node:16006) UnhandledPromiseRejectionWarning: MongoError: unknown top level operator: $disableSoftDelete
    at Function.MongoError.create (/home/robert/Development/human-connection/api/node_modules/mongodb-core/lib/error.js:31:11)
    at /home/robert/Development/human-connection/api/node_modules/mongodb-core/lib/connection/pool.js:497:72
    at authenticateStragglers (/home/robert/Development/human-connection/api/node_modules/mongodb-core/lib/connection/pool.js:443:16)
    at Connection.messageHandler (/home/robert/Development/human-connection/api/node_modules/mongodb-core/lib/connection/pool.js:477:5)
    at Socket.<anonymous> (/home/robert/Development/human-connection/api/node_modules/mongodb-core/lib/connection/connection.js:333:22)
    at Socket.emit (events.js:182:13)
    at Socket.EventEmitter.emit (domain.js:442:20)
    at addChunk (_stream_readable.js:283:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at TCP.onread (net.js:635:20)
```
Read this potentially relevant thread: https://github.com/feathers-plus/feathers-hooks-common/issues/128#issuecomment-405309822

Is there a proper way to `service.find` all contributions including soft-deleted contributions?

Shall we directly access `mongoose`?

**Comment by roschaefer** [comment]
@visualjerk No way, I overlooked the correct syntax? Got it from here: https://feathers-plus.github.io/v1/feathers-hooks-common/#softdelete

Actually the only test failing is the one I introduced, I will check if the test or the implementation is to blame..

**Comment by roschaefer** [comment]
No I get the `slug already exists error` again.



slug already exists error

**Comment by roschaefer** [comment]
@visualjerk I believe `$disableSoftDelete` belongs indeed in `query`...

**Comment by visualjerk** [comment]
@roschaefer Your are right. The problem is that it belongs in query, but throws an error, when put there. It looks like $disableSoftDelete can be considered unsupported. Moreover it is indeed a security issue, as stated in the linked issue above.

To get around this I introduced a new hook 'includeAll'. It simply deletes the parameters softDelete and excludeDisabled add to the query.

File: server/hooks/include-all.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module.exports = () => hook => {
+  if (!hook.params._includeAll) {
+    return hook;
+  }
+  if (hook.params.query.deleted) {
+    delete hook.params.query.deleted;
+  }
+  if (hook.params.query.isEnabled) {
+    delete hook.params.query.isEnabled;
+  }
```
[the remark this answers is not in the document]
Comment (visualjerk) [reply server/hooks/include-all.js]: Yes, it reverts `softDelete()` for `service.find()`

`softDelete()` adds this to the query: `query.deleted = { $ne: true };`

`includeAll()` just deletes this parameter and must be used after `softDelete()`.

As `softDelete()` is only allowed as a `all`-hook and `includeAll()` is only useful as a `find`-hook the proper order is given automatically. (`all`-hooks are executed before any specific ones)

**Pull request merged**